design/keycloak.x/configuration.md

```diff
@@ -0,0 +1,159 @@
+would like to achieve something similar to https://quarkus.io/guides/all-config.
+
+In the documentation the property key used in the properties file should be used. It should include a description, the 
+type and the default value.
+
+We would also like the ability to list properties on the command-line by running `bin/kc.sh --list-config-options`.
+
+For SPIs and Providers we should extend the ProviderFactory to add an option method to return information on what
+properties it supports.
+
+## Config Options
+
+The aim is to start small with what config options we provide, and extend when there is a need. Great care should be 
+taken in reducing the amount of configuration needed, where opinionated/best-practice behaviour should be favoured, with
+high-level and simple configuration options introduced only when needed.
+
+Below is a list of config by category we should introduce initially:
+
+### HTTP
+
+Keycloak.X will not come with the capability to generate a self-signed certificate. Instead, Keycloak by default will
+complain if a certificate has not been configured, but will allow explicitly disabling SSL for development purposes.
+
+* http.mode - Options `ssl` (default), `unsecure`  
+* http.port
+* http.ssl-port
+* http.ssl.certificate.file
+* http.ssl.certificate.key-file
+* http.ssl.certificate.key-store-file
+* http.ssl.certificate.key-store-file-type
+* http.ssl.certificate.key-store-password
+* http.ssl.certificate.trust-store-file
+* http.ssl.certificate.trust-store-file-type
+* http.ssl.certificate.trust-store-password
+* http.ssl.cipher-suites
+* http.ssl.protocols
+* http.ssl.client-auth
+
+### Proxy Mode
+
+The proxy mode will allow an easy way to configure if Keycloak is hosted behind a reverse-proxy or not, including how
+SSL is terminated.
+
+* proxy.mode - Options are `none` (default), `reencrypt`, `edge` and `passthrough`.
+
+Setting `proxy.mode` will under the covers configure `http.mode` as well as `quarkus.http.proxy-address-forwarding`.
+
+## Database
+
+The aim is that a user should not need to understand Hibernate, JDBC drivers, etc. to be able to configure the database.
+As such we should introduce a high-level option `db.vendor` instead of requiring users to set JDBC driver and Hibernate
+dialects.
+
+We will support either `db.url` which will support a JDBC URL, or using separate properties to build-up the URL behind the
+scenes.
+
+* db.vendor
+* db.url 
+* db.hostname
+* db.port - defaults to the default for the configured vendor 
+* db.schema
+* db.database
+* db.username
+* db.password
+* db.pool.initialSize
+* db.pool.minSize - default to 0
+* db.pool.maxSize - default to 100
+
+## Caches (Infinispan and JGroups)
+
+As we aim to not use an embedded Infinispan server in the future, configuration of caches will not for now follow the
+standard way of configuring Keycloak. Instead there will be a separate XML file for configuring this (`conf/cluster.xml`).
+
+## Logging
+
+TBD
+
+## Metrics
```

Comment:
I would say we need a separate document for this one so that we can come up with a nice solution that leverages MP Metrics that make easier setup richer and meaningful Grafana dashboard from the default metrics we provide, for instance.

Reply:
Yep, hence it's TBD